File: src/lib/watchdog.ts
import { invoke } from '@tauri-apps/api/core';

/**
 * HPC job tracking.
 *
 * Operon used to run a bash daemon (`operon-watchdog.sh`) on the cluster's LOGIN
 * node that polled SLURM forever so job state stayed fresh while the app was
 * closed. HPC sites reap exactly that kind of process — UCI RCIC terminated ours
 * and emailed the account owner — and it was re-bootstrapped on every SSH
 * connect, so the kill/restart cycle never ended.
 *
 * There is no daemon now. The panel asks the scheduler directly, only while it
 * is open: {@link listClusterJobs} for state, {@link readJobLogTail} for logs.
 * For "tell me when it finishes even though Operon is closed", set the optional
 * notification email in the SSH profile's server settings — SLURM mails you
 * itself, which needs nothing of ours to be running.
 */

/** A job as the Jobs panel sees it — a live `squeue` row or an `sacct` record. */
export interface ClusterJob {
  job_id: string;
  name: string;
  state: string;
  partition: string;
  /** Human elapsed from squeue ("1:23:45"); empty for historical rows. */
  elapsed: string;
  /** Seconds from sacct; 0 when unknown. */
  elapsed_seconds: number;
  /** squeue's NODELIST(REASON) — why it's pending, or where it runs. */
  reason: string;
  /** sacct ExitCode ("0:0"); empty while running. */
  exit_code: string;
  /** End time exactly as sacct reported it, in the cluster's local time
   *  ("2026-08-13T07:20:03"); empty while running or unknown. Deliberately not
   *  an epoch — converting cost a `date` fork per row on the login node. */
  ended_at: string;
  source: 'squeue' | 'sacct';
}

export interface ClusterJobsResult {
  jobs: ClusterJob[];
  /** False when the remote shell could resolve neither $USER nor `id -un`, so
   *  the query was skipped rather than run unfiltered (which would list the
   *  whole cluster, each row with a working Cancel button). */
  user_resolved: boolean;
  /**
   * False when the cluster has no usable SLURM accounting. Without `sacct` a job
   * simply vanishes from `squeue` when it ends and there is no record left to
   * read, so the panel must say so rather than imply the job never existed.
   */
  accounting: boolean;
}

/** SLURM states that mean the job is over. Mirrors `is_terminal_state` in slurm.rs. */
const TERMINAL = new Set([
  'COMPLETED',
  'FAILED',
  'TIMEOUT',
  'OUT_OF_MEMORY',
  'NODE_FAIL',
  'BOOT_FAIL',
  'DEADLINE',
  'PREEMPTED',
  'REVOKED',
  'SPECIAL_EXIT',
]);

export function isTerminalState(state: string): boolean {
  const s = (state || '').trim().toUpperCase().split(/\s+/)[0] || '';
  return TERMINAL.has(s) || s.startsWith('CANCELLED');
}

/**
 * Query the cluster for this user's jobs — live plus recent history — in one
 * round-trip.
 *
 * @param since       SLURM time expression; defaults to `now-7days`.
 */
export async function listClusterJobs(
  profileId: string,
  since?: string,
): Promise<ClusterJobsResult> {
  return invoke('list_cluster_jobs', { profileId, since: since || null });
}

/** Read the tail of a job's stdout log on demand (no `tail -f` left running). */
export async function readJobLogTail(
  profileId: string,
  jobId: string,
  logPath?: string | null,
  lines?: number,
): Promise<string> {
  return invoke('read_job_log_tail', {
    profileId,
    jobId,
    logPath: logPath || null,
    lines: lines ?? null,
  });
}


export interface LegacyCleanupResult {
  /** True only when a leftover daemon or its files were actually found. */
  removed: boolean;
  details: string;
}

/**
 * Kill and delete any `operon-watchdog.sh` daemon left over from an older
 * Operon. Idempotent and safe on hosts that never had one.
 *
 * Needed because upgrading does not stop a daemon that is already running on the
 * user's login node — without this sweep they would keep receiving kill notices
 * from their site long after the feature was removed.
 */
export async function cleanupLegacyWatchdog(profileId: string): Promise<LegacyCleanupResult> {
  return invoke('cleanup_legacy_watchdog', { profileId });
}

// ── auto-register helper ─────────────────────────────────────────────────

const SBATCH_RE = /Submitted batch job\s+(\d+)/g;

/**
 * Scan a chunk of terminal output for `Submitted batch job NNNN` and return
 * any job ids found. Caller dedupes and calls `registerSlurmJob` for each hit.
 */
export function parseSbatchIds(text: string): string[] {
  // Global match, and over the whole text rather than line-by-line: agent output
  // arrives as NDJSON, so several `Submitted batch job` strings can share one
  // physical line (their newlines are escaped inside the JSON). The old
  // first-match-per-line scan silently dropped all but one.
  const ids: string[] = [];
  for (const m of text.matchAll(SBATCH_RE)) ids.push(m[1]);
  return ids;
}

File: src-tauri/src/commands/watchdog.rs
//! HPC job tracking — scheduler detection and legacy-daemon cleanup.
//!
//! ── What used to be here ──
//! Operon shipped a job "watchdog": `scripts/operon-watchdog.sh`, uploaded to
//! the remote host and launched detached (tmux, or `nohup … & disown`) so it
//! outlived the app. It polled SLURM every 30s forever from the **login node**.
//!
//! That was the wrong design, and it had real consequences:
//!
//!   * HPC sites reap long-lived login-node processes. UCI RCIC terminated ours
//!     by name — its matcher hit the substring `watch` in `operon-watchdog.sh` —
//!     and emailed the account owner every time. Because the daemon was
//!     re-bootstrapped on every SSH connect, the kill/restart cycle repeated
//!     indefinitely.
//!   * The loop had no exit condition: no wall-clock cap, no idle timeout, no
//!     exit when the watchlist emptied. Its `squeue` self-discovery ran *before*
//!     the empty-watchlist guard, so it queried the scheduler every 30s even
//!     with zero jobs, and the watchlist only ever grew.
//!   * Its one genuinely stateful feature — auto-resubmit on TIMEOUT/OOM — was
//!     broken: `slurm_resubmit` re-ran the *identical* script. The
//!     `on_timeout_walltime_mult` / `on_oom_mem_mult` knobs in the UI were never
//!     applied by anything, so a TIMEOUT was resubmitted with the same walltime
//!     and timed out again, burning the allocation to reproduce the failure.
//!
//! ── What replaced it ──
//! Nothing resident. The Jobs panel needs *data*, not a process:
//!
//!   * live + historical job state → `slurm::list_cluster_jobs` (`squeue` +
//!     `sacct` in one round-trip, issued only while the panel is open)
//!   * job logs → `slurm::read_job_log_tail` (on demand, no `tail -f`)
//!   * completion while Operon is closed → SLURM's own `--mail-user`, from the
//!     profile's optional `notify_email` server config
//!   * session↔job attribution → `job_notify`'s LOCAL registry
//!
//! Job tracking now leaves nothing running on the cluster. Note this is not the
//! same as "Operon never has a login-node process": an active HPC *chat* session
//! still opens a `tail -f` there to stream the agent's output (`claude.rs`).
//! That one is bounded (12h cap + PPID watcher), dies with the session, and is
//! reapable — its children reset the HUP/PIPE traps the parent ignores, so it
//! can't survive as an orphan the way the daemon did.

use serde::Serialize;

use super::ssh::{ssh_exec, SSHManager};

// ─── Legacy daemon cleanup ───────────────────────────────────────────────

/// Result of sweeping a host for the retired watchdog daemon.
#[derive(Debug, Clone, Serialize)]
pub struct LegacyCleanupResult {
    /// True when something was actually found and removed — lets the UI stay
    /// silent on the overwhelmingly common "nothing there" case.
    pub removed: bool,
    pub details: String,
}

/// Kill and delete any leftover `operon-watchdog.sh` daemon and its remote
/// state. Idempotent, one-shot, and safe to run on a host that never had one.
///
/// This exists for users upgrading from a build that installed the daemon: it
/// is still running on their login node right now, and nothing in the new code
/// path would ever stop it. Without this they would keep getting kill notices
/// from their site long after the feature was removed.
#[tauri::command]
pub async fn cleanup_legacy_watchdog(
    ssh_state: tauri::State<'_, SSHManager>,
    profile_id: String,
) -> Result<LegacyCleanupResult, String> {
    let profile = ssh_state
        .profiles
        .lock()
        .map_err(|e| e.to_string())?
        .iter()
        .find(|p| p.id == profile_id)
        .cloned()
        .ok_or_else(|| format!("SSH profile {} not found", profile_id))?;

    // Preflight so an MFA cluster returns "open an SSH terminal and complete
    // Duo" rather than an opaque BatchMode auth failure — every path here is
    // non-interactive and can only succeed by riding a live ControlMaster.
    super::ssh::ensure_live_connection(&profile)?;

    // `operon-[w]atchdog` throughout: the bracket makes the pattern unable to
    // match the shell running this very script, whose command line contains the
    // literal "[w]" that the character class does not accept. Without it,
    // pkill would kill the cleanup itself before it finished.
    let script = r#"set -u
_found=0
if tmux has-session -t operon-watchdog 2>/dev/null; then
  _found=1
  tmux kill-session -t operon-watchdog 2>/dev/null
fi
# The pid file is NOT trusted on its own. The population this sweep targets is
# exactly the one holding a STALE pidfile — the site reaper killed the daemon and
# left the file behind — and login nodes have long uptimes, so that pid has very
# likely been recycled onto something else of the user's: a tmux server, an
# editor, an interactive srun shell. `kill -0` only proves "exists and is mine",
# which is not the same question. Verify the process is actually the watchdog
# before signalling it, and ignore a non-numeric file.
if [ -f "$HOME/.operon/watchdog.pid" ]; then
  _p=$(cat "$HOME/.operon/watchdog.pid" 2>/dev/null)
  case "$_p" in
    ''|*[!0-9]*) ;;
    *)
      if ps -p "$_p" -o args= 2>/dev/null | grep -q 'operon-[w]atchdog'; then
        _found=1
        kill "$_p" 2>/dev/null
      fi
      ;;
  esac
fi
if pgrep -f 'operon-[w]atchdog\.sh' >/dev/null 2>&1; then
  _found=1
  pkill -f 'operon-[w]atchdog\.sh' 2>/dev/null
fi
[ -f "$HOME/.operon/operon-watchdog.sh" ] && _found=1
rm -f "$HOME/.operon/operon-watchdog.sh" \
      "$HOME/.operon/watchdog.pid" \
      "$HOME/.operon/watchdog.log" \
      "$HOME/.operon/watchlist" \
      "$HOME/.operon/watchlist.tmp."* \
      "$HOME/.operon/policy.json" 2>/dev/null
rm -rf "$HOME/.operon/jobs" 2>/dev/null
# Leave ~/.operon itself — other Operon state may live there. Remove it only if
# our removals emptied it.
rmdir "$HOME/.operon" 2>/dev/null
echo "removed=$_found"
"#;

    let out = ssh_exec(&profile, script).map_err(|e| format!("cleanup failed: {}", e))?;
    let removed = out.lines().any(|l| l.trim() == "removed=1");
    Ok(LegacyCleanupResult {
        removed,
        details: out.trim().to_string(),
    })
}
